### PR TITLE
Merge DMOJ into master

### PR DESCRIPTION
**Pull request by fluix-dev**

### CODE REVIEW

File: judge/models/contest.py [hunk 1]
```diff
@@ -338,16 +344,19 @@ def access_check(self, user):
...
-            if self.is_private and not self.is_organization_private:
-                if in_users:
-                    return
-                raise self.PrivateContest()
+        if self.is_private and not (self.is_organization_private or self.is_private_viewable):
+            if in_users:
+                return
+            raise self.PrivateContest()
 
             if self.is_private and (self.is_organization_private or self.is_private_viewable):
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: Indentation is still the wrong level. Also you're comparing `self.is_private and (self.is_organization_private or self.is_private_viewable)` twice? (Once here and once on line 356).

File: judge/models/contest.py [hunk 2]
```diff
@@ -218,6 +218,12 @@ def can_see_full_scoreboard(self, user):
             return True
         if not self.show_scoreboard:
             return False
+        if not self.is_visible:
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: Code doesn't belong here (Does it even make sense to be here? Also, a lot of this code is already in `is_accessible_by`).

**Review by Ninjaclasher — CHANGES_REQUESTED** [review]

File: judge/models/contest.py (hunk 1 above)
Comment (fluix-dev) [reply judge/models/contest.py]: This line and the one on 356 are not the same.

File: judge/models/contest.py [hunk 3]
```diff
@@ -218,6 +218,10 @@ def can_see_full_scoreboard(self, user):
             return True
         if not self.show_scoreboard:
             return False
+        if self.start_time is not None and self.start_time > timezone.now():
+            return False
+        if self.hide_scoreboard and not self.is_in_contest(user) and self.end_time > timezone.now():
+            return False
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: Not just that one if, _all_ of this.

**Review by Ninjaclasher — APPROVED** [review]

**Pull request merged**